store/src/nullifier_tree.rs

```diff
@@ -0,0 +1,110 @@
+use miden_objects::{
+    crypto::{
+        hash::rpo::RpoDigest,
+        merkle::{Smt, SmtProof},
+    },
+    notes::Nullifier,
+    Felt, FieldElement, Word, EMPTY_WORD,
+};
+
+use crate::{errors::NullifierTreeError, types::BlockNumber};
+
+/// Nullifier SMT.
+#[derive(Debug, Clone)]
+pub struct NullifierTree(Smt);
+
+impl NullifierTree {
+    /// Construct new nullifier tree from list of items.
+    pub fn with_entries(
+        entries: impl IntoIterator<Item = (Nullifier, BlockNumber)>
+    ) -> Result<Self, NullifierTreeError> {
+        let leaves = entries.into_iter().map(|(nullifier, block_num)| {
+            (nullifier.inner(), Self::block_num_to_leaf_value(block_num))
+        });
+
+        let inner = Smt::with_entries(leaves)?;
+
+        Ok(Self(inner))
+    }
+
+    /// Get SMT root.
+    pub fn root(&self) -> RpoDigest {
+        self.0.root()
+    }
+
+    /// Returns an opening of the leaf associated with the given nullifier.
+    pub fn open(
+        &self,
+        nullifier: &Nullifier,
+    ) -> SmtProof {
+        self.0.open(&nullifier.inner())
+    }
+
+    /// Inserts block number in which nullifier was consumed.
+    pub fn insert(
+        &mut self,
+        nullifier: &Nullifier,
+        block_num: BlockNumber,
+    ) -> Result<(), NullifierTreeError> {
+        let value = self.0.insert(nullifier.inner(), Self::block_num_to_leaf_value(block_num));
+        if value == EMPTY_WORD {
+            return Ok(());
+        }
+
+        Err(NullifierTreeError::NullifierAlreadyExists {
+            nullifier: *nullifier,
+            block_num: Self::leaf_value_to_block_num(value),
+        })
+    }
+
+    /// Returns block number stored for the given nullifier or `None` if the nullifier wasn't
+    /// consumed.
+    pub fn get_block_num(
+        &self,
+        nullifier: &Nullifier,
+    ) -> Option<BlockNumber> {
+        let value = self.0.get_value(&nullifier.inner());
+        if value == EMPTY_WORD {
+            return None;
+        }
```

Comment:
This feels like a design flaw in `SparseMerkleTree`. We are required to encode the lack of a value as an `EMPTY_WORD`, but that implementation detail is leaking, and now on higher level abstractions we are having to special case it.

https://github.com/0xPolygonMiden/crypto/issues/289

Reply:
Building on https://github.com/0xPolygonMiden/crypto/issues/289#issuecomment-1997266377, the correct check here would be `if value == Smt::EMPTY_VALUE`, since `Smt::EMPTY_VALUE` is part of the public API, but the fact that `Smt::EMPTY_VALUE == EMPTY_WORD` is an implementation detail.

Reply:
Yes, let's change this to `if value == Smt::EMPTY_VALUE`.